detox/src/devices/allocation/drivers/android/genycloud/GenyAllocDriver.js

```diff
@@ -28,21 +29,24 @@ class GenyAllocDriver extends AllocationDriverBase {
     const recipe = await this._recipeQuerying.getRecipeFromQuery(deviceQuery);
     this._assertRecipe(deviceQuery, recipe);
 
-    const allocResult = await this._instanceAllocationHelper.allocateDevice(recipe);
-    let { instance, isNew } = allocResult;
+    const { instance, isNew } = await this._instanceAllocationHelper.allocateDevice(recipe);
+    this._launchInfo[instance.uuid] = { isNew };
+    return new GenycloudEmulatorCookie(instance);
+  }
 
-    try {
-      instance = await this._instanceLauncher.launch(instance, isNew);
-    } catch (e) {
-      await this._instanceAllocationHelper.deallocateDevice(instance.uuid);
-      throw e;
-    }
-    const { adbName } = instance;
+  /**
+   * @param {GenycloudEmulatorCookie} cookie
+   * @returns {Promise<void>}
+   */
+  async postAllocate(cookie) {
+    const { instance } = cookie;
+    const { isNew } = this._launchInfo[instance.uuid];
+    const readyInstance = cookie.instance = await this._instanceLauncher.launch(instance, isNew);
 
+    const { adbName } = readyInstance;
     await this._adb.disableAndroidAnimations(adbName);
```

Comment:
@d4vidi, you might review this pull request post-factum and wonder why I "kinda applied an overkill", but there was a real reason.

In case with that Genymotion bug (referenced in this PR), the problem was that it DID NOT throw any errors, it was just stuck for minutes-minutes (eternity). So the deallocation just never happened. This `try-catch` was simply not enough to handle this kind of errors.

`init()` - `cleanup()` cycles in DetoxWorker are controlled much tighter (we have `yield`, CAF cancellation, and Jest environment timers which correctly manage even the messiest situations), so I had no better choice than to go up one level and join the YIELD* party in init.

Reply:
And if you ever wonder why I didn't try to fix this solely via #4006 – well, in my opinion that's a too local and specific fix which conceptually doesn't guarantee that this situation won't happen again. Splitting code into pre-cookie and post-cookie is much safer when it comes to deallocating via cookie, and that provides a generic solution to this issue.